### PR TITLE
Default HTTPS for fed sharing & config option for allowing fallback

### PR DESCRIPTION
**Pull request by tomneedham**
## Description
Only uses HTTPS for federated share communication - unless admins explicitly allow HTTP (useful for testing as well)

## Related Issue
Potential for users to setup shares via HTTP with unsecure servers without the admin knowing.

## Motivation and Context
MITM

## How Has This Been Tested?
not yet

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: lib/private/Share/Share.php [hunk 1]
```diff
@@ -2666,6 +2666,7 @@ public static function removeProtocolFromUrl($url) {
 	 * @return array
 	 */
 	private static function tryHttpPostToShareEndpoint($remoteDomain, $urlSuffix, array $fields) {
+		$allowHttpFallback = \OC::$server->getConfig()->getSystemValue('sharing.federation.allowHttpFallback', 'no') === 'yes';
```
Comment (mmattel) [inline lib/private/Share/Share.php]: In ``config.sample.php`` above you have as values ``true/false`` but here you use ``yes/no``.
Suggesting to use ``true/false`` homogeniously.

Comment (tomneedham) [reply lib/private/Share/Share.php]: Thanks @mmattel for the spot. I was originally using appconfig but concluded on config.php so we can use booleans here. Updated the PR to reflect this.

**Comment by PVince81** [comment]
Triple CI failure due to bad luck.

Please rebase and solve conflict and be more lucky

**Comment by tomneedham** [comment]
rebased, squashed and resolved conflict - let's wait for CI

**Comment by PVince81** [comment]
```
12:49:08 1) Test\Share\ShareTest::testRemoteShareUrlCalls with data set #0 ('admin@localhost', 'localhost', false, false)
12:49:08 Exception: Sharing test.txt failed, could not find admin@localhost, check spelling and server availability.
12:49:08 
12:49:08 /var/lib/jenkins/workspace/owncloud-core_core_PR-30198-OBFZEBOE7XKCZTFDCCEZHK2P56TXDW23OPZENHUDKP2Z245D372Q/lib/private/Share/Share.php:919
12:49:08 /var/lib/jenkins/workspace/owncloud-core_core_PR-30198-OBFZEBOE7XKCZTFDCCEZHK2P56TXDW23OPZENHUDKP2Z245D372Q/lib/public/Share.php:267
12:49:08 /var/lib/jenkins/workspace/owncloud-core_core_PR-30198-OBFZEBOE7XKCZTFDCCEZHK2P56TXDW23OPZENHUDKP2Z245D372Q/tests/lib/Share/ShareTest.php:973
12:49:08 
12:49:08 2) Test\Share\ShareTest::testRemoteShareUrlCalls with data set #1 ('admin@https://localhost', 'localhost', false, false)
12:49:08 Exception: Sharing test.txt failed, could not find admin@https://localhost, check spelling and server availability.
12:49:08 
12:49:08 /var/lib/jenkins/workspace/owncloud-core_core_PR-30198-OBFZEBOE7XKCZTFDCCEZHK2P56TXDW23OPZENHUDKP2Z245D372Q/lib/private/Share/Share.php:919
12:49:08 /var/lib/jenkins/workspace/owncloud-core_core_PR-30198-OBFZEBOE7XKCZTFDCCEZHK2P56TXDW23OPZENHUDKP2Z245D372Q/lib/public/Share.php:267
12:49:08 /var/lib/jenkins/workspace/owncloud-core_core_PR-30198-OBFZEBOE7XKCZTFDCCEZHK2P56TXDW23OPZENHUDKP2Z245D372Q/tests/lib/Share/ShareTest.php:973
12:49:08 
12:49:08 3) Test\Share\ShareTest::testRemoteShareUrlCalls with data set #3 ('admin@localhost/subFolder', 'localhost/subFolder', false, false)
12:49:08 Exception: Sharing test.txt failed, could not find admin@localhost/subFolder, check spelling and server availability.
12:49:08 
12:49:08 /var/lib/jenkins/workspace/owncloud-core_core_PR-30198-OBFZEBOE7XKCZTFDCCEZHK2P56TXDW23OPZENHUDKP2Z245D372Q/lib/private/Share/Share.php:919
12:49:08 /var/lib/jenkins/workspace/owncloud-core_core_PR-30198-OBFZEBOE7XKCZTFDCCEZHK2P56TXDW23OPZENHUDKP2Z245D372Q/lib/public/Share.php:267
12:49:08 /var/lib/jenkins/workspace/owncloud-core_core_PR-30198-OBFZEBOE7XKCZTFDCCEZHK2P56TXDW23OPZENHUDKP2Z245D372Q/tests/lib/Share/ShareTest.php:973
12:49:08 
12:49:08 4) Test\Share\ShareTest::testRemoteShareUrlCalls with data set #4 ('admin@localhost/subFolder', 'localhost/subFolder', false, false)
12:49:08 Exception: Sharing test.txt failed, could not find admin@localhost/subFolder, check spelling and server availability.
12:49:08 
12:49:08 /var/lib/jenkins/workspace/owncloud-core_core_PR-30198-OBFZEBOE7XKCZTFDCCEZHK2P56TXDW23OPZENHUDKP2Z245D372Q/lib/private/Share/Share.php:919
12:49:08 /var/lib/jenkins/workspace/owncloud-core_core_PR-30198-OBFZEBOE7XKCZTFDCCEZHK2P56TXDW23OPZENHUDKP2Z245D372Q/lib/public/Share.php:267
12:49:08 /var/lib/jenkins/workspace/owncloud-core_core_PR-30198-OBFZEBOE7XKCZTFDCCEZHK2P56TXDW23OPZENHUDKP2Z245D372Q/tests/lib/Share/ShareTest.php:973
```

**Comment by tomneedham** [comment]
Fixed the failing tests - runs fine locally.

**Comment by PVince81** [comment]
hmm there are still failures:
```
1) Test\Share\ShareTest::testRemoteShareUrlCalls with data set #0 ('admin@localhost', 'localhost', false, false)
Exception: Sharing test.txt failed, could not find admin@localhost, check spelling and server availability.

/drone/src/lib/private/Share/Share.php:919
/drone/src/lib/public/Share.php:267
/drone/src/tests/lib/Share/ShareTest.php:973
```

**Comment by PVince81** [comment]
Running all tests locally with PHP 7.1 and sqlite does not produce this error. I even shut down my local web server in case it was trying to connect there...

Mystery...

Let's rebase this anyway as we have library updates on master

**Comment by PVince81** [comment]
@ownclouders rebase

**Comment by ownclouders** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Comment by ownclouders** [comment]
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

**Comment by PVince81** [comment]
tests still pass locally...

**Comment by PVince81** [comment]
@tom please backport

**Comment by phil-davis** [comment]
Backport stable10 #30646

**Comment by PVince81** [comment]
Not working, raised here https://github.com/owncloud/core/issues/31194

**Pull request merged**